akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -283,14 +291,21 @@ private[akka] object StreamLayout {
 
     override def toString =
       s"""
-        | Module: ${this.attributes.nameOption.getOrElse("unnamed")}
-        | Modules: ${subModules.toSeq.map(m ⇒ "   " + m.attributes.nameOption.getOrElse(m.getClass.getName)).mkString("\n")}
+        | Module: ${this.attributes.nameOrDefault("unnamed")}
+        | Modules: ${subModules.iterator.map(m ⇒ "   " + m.attributes.nameOrDefault(m.getClass.getName)).mkString("\n")}
         | Downstreams:
-        | ${downstreams.map { case (in, out) ⇒ s"   $in -> $out" }.mkString("\n")}
+        | ${downstreams.iterator.map { case (in, out) ⇒ s"   $in -> $out" }.mkString("\n")}
         | Upstreams:
-        | ${upstreams.map { case (out, in) ⇒ s"   $out -> $in" }.mkString("\n")}
+        | ${upstreams.iterator.map { case (out, in) ⇒ s"   $out -> $in" }.mkString("\n")}
       """.stripMargin
 
+    override def describe: String =
+      attributes.nameOrDefault(null) match {
+        case null ⇒ "(" + subModules.iterator.map(_.describe).mkString(" ~> ") + ")"
```

Comment:
One of the dragons have been eliminated (which forced me to not merge it that time) and that was copying of imported graphs. Now we have a marker CopiedModule instead, and this helped to simplify the wiring code. 

The idea roughly:
 - create a LinearComposite case class that extends Module and which only have at most one external (unwired) output and at most one external input port.
 - override growConnect to handle the case when the Module is wired linearly
 - once non-linear wiring is detected, convert the thing to the existing CompositeModule (since no longer possible to represent the thing as a chain, it is now a graph)
 - wrap() should return a LinearComposite, so we always assume that a graph is linear, until we detect its not.

At least this is how I did it. External users (the user DSLs) should not be aware of the existence of this internal optimization. The MaterializerSession can optionally make use of this extra knowledge to make materialization faster, but this is a nice to have.